src/pip/_internal/locations.py

```diff
@@ -138,7 +138,7 @@ def distutils_scheme(
 
     if running_under_virtualenv():
         scheme['headers'] = os.path.join(
-            sys.prefix,
+            i.prefix,
```

Comment:
This whole block is a workaround for the normal `include/pythonX.Y` being a symlink in virtualenv on Unix (description [here](https://www.python.org/dev/peps/pep-0405/#include-files)). Whether this is something we want to use here will depend on what the user wants the provided `--prefix` to look like:

* normal - in which case we shouldn't execute this block at all
* virtualenv - in which case the current approach looks good to me

Tooling-wise it may not matter if every tool out there already looks for headers in the normal places and our non-standardized path. If that's true then, while environment dependent, this proposed approach is better than what we have now and we'd still be able to change it later.

Reply:
`distutils` automatically does the right thing if pip is not running in a virtual environment:

```pycon
>>> import sys, pip
>>> pip.__version__
'19.2.3'
>>> sys.prefix
'/Users/uranusjr/Library/PythonUp/versions/3.8'
>>> from pip._internal.locations import distutils_scheme
>>> distutils_scheme('', prefix=None)['headers']
'/Users/uranusjr/Library/PythonUp/versions/3.8/include/python3.8/UNKNOWN'
>>> distutils_scheme('', prefix='foo')['headers']
'foo/include/python3.8/UNKNOWN'
```

Only the virtualenv case is currently breaks with `--prefix`.

Reply:
I added a sentence in the news fragment to make this clearer.

Reply:
I follow. If pip isn't running in a virtualenv then this block doesn't matter. I'm just saying it seems odd that if pip happens to be executing from a virtual environment it would impose its custom virtualenv header path logic on a `--prefix` provided by the user, which may have nothing to do with virtualenv. I think it's probably OK, and it's better than what we're doing now, but it would help to understand what the expectations of users and build tools are.

Reply:
On second thought, it would probably just require introducing some flags and incrementally deprecating and adopting the new functionality, so I don't see a problem taking this step now.

Reply:
Yeah, I don’t get why pip uses a custom header path logic here either. `distutils` by defaults returns `{VIRTUAL_ENV}/include/python{X}.{Y}/{dist_name}`, which seems as good as `{VIRTUAL_ENV}/include/python{X}.{Y}/site/{dist_name}` to me. Unfortunately this logic is from a [time](https://github.com/pypa/pip/commit/9c271baafe3aea7ee8363a108c91a6f948415a62) when nobody cares to record any reasoning on anything pip does.

I [[envision]] somethime in the future pip should deprecate and replace the whole distutils-based logic with `sysconfig`.

Reply:
I have a vague recollection that historically the headers weren't included in the "scheme" layout. That may have changed at the point when core Python gained `venv`, and could no longer ignore questions of how virtual environments should work 🙂 But I agree, we should really be deferring everything about "installation schemes" to sysconfig (where applicable, things like `--target` should have handling that's `sysconfig`-compatible, but sysconfig probably won't cover that sort of case directly).

I suspect that switching fully to sysconfig will only be possible once the legacy direct install path has been finally dropped...

Reply:
I linked a PEP [above](https://www.python.org/dev/peps/pep-0405/#include-files) that contains the original justification.